client/grpc_query.go

```diff
@@ -17,12 +20,12 @@ var _ gogogrpc.ClientConn = Context{}
 var protoCodec = encoding.GetCodec(proto.Name)
 
 // Invoke implements the grpc ClientConn.Invoke method
-func (ctx Context) Invoke(_ gocontext.Context, method string, args, reply interface{}, _ ...grpc.CallOption) error {
+func (ctx Context) Invoke(grpcCtx gocontext.Context, method string, args, reply interface{}, opts ...grpc.CallOption) error {
 	reqBz, err := protoCodec.Marshal(args)
 	if err != nil {
 		return err
 	}
-	resBz, _, err := ctx.QueryWithData(method, reqBz)
+	resBz, height, err := ctx.QueryWithData(method, reqBz)
```

Comment:
we should use `QueryABCI` instead of `QueryWithData` so that we can include the proofs in the header metadata too

Reply:
If i'm not mistaken, QueryABCI is called under the hood:

- first https://github.com/cosmos/cosmos-sdk/blob/master/client/query.go#L37
- then https://github.com/cosmos/cosmos-sdk/blob/master/client/query.go#L109

Reply:
yes, I was referring to use the public QueryABCI, so that we can add the proofs to the metadata at a later stage:

```go
func (ctx Context) Invoke(grpcCtx gocontext.Context, method string, args, reply interface{}, opts ...grpc.CallOption) error {
	reqBz, err := protoCodec.Marshal(args)
	if err != nil {
		return err
	}
	req := abci.RequestQuery{
		Path: method,
		Data: reqBz,
	}

	res, err := ctx.QueryABCI(req)
	if err != nil {
		return err
	}

	err = protoCodec.Unmarshal(res.Value, reply)
	if err != nil {
		return err
	}

	// Create header metadata. For now the headers contain:
	// - block height
	// We then parse all the call options, if the call option is a
	// HeaderCallOption, then we manually set the value of that header to the
	// metadata.
	md := metadata.Pairs(baseapp.GRPCBlockHeightHeader, strconv.FormatInt(res.Height, 10))
	for _, callOpt := range opts {
		header, ok := callOpt.(grpc.HeaderCallOption)
		if !ok {
			continue
		}

		*header.HeaderAddr = md
	}

	if ctx.InterfaceRegistry != nil {
		return types.UnpackInterfaces(reply, ctx.InterfaceRegistry)
	}

	return nil
}
```